src/treeffuser/_base_tabular_diffusion.py

```diff
@@ -97,41 +99,137 @@ def get_new_score_model(self) -> ScoreModel:
         Return the score model.
         """
 
-    def _validate_data(
+    def _preprocess_and_validate_data(
         self,
-        X: Optional[ndarray] = None,
-        y: Optional[ndarray] = None,
+        X: Optional[Union[ndarray, pd.DataFrame]] = None,
+        y: Optional[Union[ndarray, pd.Series]] = None,
+        cat_idx: Optional[List[int]] = None,
         validate_X: bool = True,
         validate_y: bool = True,
+        reset: bool = False,
     ) -> Tuple[
-        Optional[Float[ndarray, "batch x_dim"]], Optional[Float[ndarray, "batch y_dim"]]
+        Optional[Float[ndarray, "batch x_dim"]],
+        Optional[Float[ndarray, "batch y_dim"]],
+        List[int],
     ]:
-        """Reshape X and/or y to adhere to the (batch, n_dim) convention, and cast them as flows."""
-        if validate_X and X is not None:
+        """
+        If X is a DataFrame:
+            - convert it to numpy array and auto-detect categorical columns.
+            - store the column names at fit time and ensure they are the same at predict time.
+        If X is a numpy array:
+            Reshape X and/or y to adhere to the (batch, n_dim) convention
+        """
+        if validate_X:
+            if X is None:
+                raise ValueError("Input data `X` cannot be None.")
+            if isinstance(X, pd.DataFrame):
+                if cat_idx is not None and cat_idx != []:
+                    raise ValueError(
+                        "Categorical indices `cat_idx` should not be provided when `X` is a DataFrame."
+                        "Instead, ensure that the categorical columns have dtype `category`."
+                        "E.g., X['column_name'] = X['column_name'].astype('category')."
+                    )
+                cat_idx = [
+                    X.columns.get_loc(col) for col in X.select_dtypes("category").columns
+                ]
+                # check dtype of columns and convert categorical columns to numerical
+                X_has_been_copied = False
```

Comment:
ntp. I feel this section here could be a function in itself cause it is a bit long and I think it could make the code more readable. But up to u!

Reply:
fixed